src/NLog/Internal/FileAppenders/BaseFileAppenderCache.cs

```diff
@@ -0,0 +1,258 @@
+            {
+                BaseFileAppender newAppender = Factory.Open(fileName, CreateFileParameters);
+
+                if (appenders[freeSpot] != null)
+                {
+                    appenders[freeSpot].Close();
+                    appenders[freeSpot] = null;
+                }
+
+                for (int j = freeSpot; j > 0; --j)
+                {
+                    appenders[j] = appenders[j - 1];
+                }
+
+                appenders[0] = newAppender;
+                appenderToWrite = newAppender;
+            }
+
+            return appenderToWrite;
+        }
+
+        public void CloseAppenders()
+        {
+            if (appenders != null)
+            {
+                for (int i = 0; i < appenders.Length; ++i)
+                {
+                    if (appenders[i] == null)
+                    {
+                        break;
+                    }
+
+                    appenders[i].Close();
+                    appenders[i] = null;
+                }
+            }
+        }
+
+        public void CloseAppenders(DateTime expireTime)
+        {
+            for (int i = 0; i < this.appenders.Length; ++i)
+            {
+                if (this.appenders[i] == null)
+                {
+                    break;
+                }
+
+                if (this.appenders[i].OpenTime < expireTime)
+                {
+                    for (int j = i; j < this.appenders.Length; ++j)
+                    {
+                        if (this.appenders[j] == null)
+                        {
+                            break;
+                        }
+
+                        this.appenders[j].Close();
+                        this.appenders[j] = null;
+                    }
+
+                    break;
+                }
+            }
+        }
+
+        // HACK: This method exposes a reference to the underlying item directly to the calling method. 
+        //    A new object should be constructed through a copy constructor. At the moment we will allow this until we
+        //    verify that there is not significant performance penalty.
+        public BaseFileAppender FindItem(string fileName)
+        {
+            foreach (BaseFileAppender appender in appenders)
+            {
+                if (appender == null)
+                {
+                    break;
+                }
+
+                if (appender.FileName == fileName)
+                {
+                    return appender;
+                }
+            }
+
+            return null;
+        }
+
+        public void FlushAppenders()
+        {
+            foreach (BaseFileAppender appender in appenders)
+            {
+                if (appender == null)
+                {
+                    break;
+                }
+
+                appender.Flush();
+            }
+        }
+
+        public bool GetFileInfo(string fileName, out DateTime lastWriteTime, out long fileLength)
+        {
```

Comment:
The out parameters declared on the abstract class BaseFileAppender. It would worth looking into this but it might considering as a more extensive refactoring.